sherpa/utils/types.py

```diff
@@ -0,0 +1,55 @@
+#
+#  Copyright (C) 2024
+#  Smithsonian Astrophysical Observatory
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+
+"""Useful types for Sherpa.
+
+This module should be considered an internal module as its contents is
+likely to change as types get added to Sherpa and the typing ecosystem
+in Python matures.
+
+"""
+
+from typing import Callable, Sequence, Union
+
+import numpy as np
+
+
+# Some of these may change to TypeAlias (once Python 3.10 is the
+# minimum) and then type/TypeAliasType once Python 3.12 is the
+# minimum.
+#
+
+# Try to be generic when using arrays as input or output. There is no
+# attempt to encode the data type or shape for ndarrays at this time.
+#
+ArrayType = Union[Sequence[float], np.ndarray]
```

Comment:
Also allow sequence of int here (I'm thinking counts in a PHA) or do you want that to be a separate type `ArrayIntType`?

Reply:
There's a couple of issues here

- technically `int` is subsumed by `float` so it's not obvious we'd gain anything by including `int` here
  - note that the numeric typing tower is surprising complex, even when you don't need to worry about IEEE 757/NaN issues, as discussed in many places, such as the recent typing discourse item: https://discuss.python.org/t/clarifying-the-float-int-complex-special-case/54018
- for PHA data, we can not say "must be an int" for the counts array (dependent axis), since we have to deal with the possibility the data may be a rate rather than a count
  - although we could enforce it for the independent axis, but there's "interesting" things to think about with that as this then changes the interface the base class promises, which can lead to worries over things like the [Liskov Sustitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle), although to be fair our existing class structure has many such places like this 

I wanted to say that this change - in particular `ArrayType`, but it holds for any of our initial typing statements - doesn't magically fix up our code. It just points to other areas that we'll have to work on in later PRs, by tweaking up the tying statements, or with code changes, or both. For example, we have many places where we send in something like `Union[float, ArryaType]` and use something like `np.iterable` to check to see whether we have an array - for example this made-up example

```
from typing import Sequence, Union

import numpy as np

ArrayType = Union[Sequence[float], np.ndarray]

def bar(xs: Union[float, ArrayType]) -> int:
    if np.iterable(xs):
        return len(xs)

    return 1
```

which current `mypy` (1.10.0) and `pyright` (1.1.368) both complain about

```
x.py:12: error: Argument 1 to "len" has incompatible type "Iterable[Any]"; expected "Sized"  [arg-type]
```

and

```
  /home/dburke/sherpa/sherpa-xspec/x.py:12:20 - error: Argument of type "Iterable[Any]" cannot be assigned to parameter "obj" of type "Sized" in function "len"
    "Iterable[Any]" is incompatible with protocol "Sized"
      "__len__" is not present (reportArgumentType)
```

Reply:
Thanks for the explanation. I always learn new things when I review your PRs!